Source/WebCore/dom/InternalObserverReduce.cpp

```diff
@@ -0,0 +1,145 @@
+#include "JSDOMPromiseDeferred.h"
+#include "Observable.h"
+#include "ReducerCallback.h"
+#include "ScriptExecutionContext.h"
+#include "SubscribeOptions.h"
+#include "Subscriber.h"
+#include "SubscriberCallback.h"
+#include <JavaScriptCore/JSCJSValueInlines.h>
+
+namespace WebCore {
+
+class InternalObserverReduce final : public InternalObserver {
+public:
+    static Ref<InternalObserverReduce> create(ScriptExecutionContext& context, Ref<AbortSignal>&& signal, Ref<ReducerCallback>&& callback, std::optional<JSC::JSValue>&& initialValue, Ref<DeferredPromise>&& promise)
+    {
+        Ref internalObserver = adoptRef(*new InternalObserverReduce(context, WTFMove(signal), WTFMove(callback), WTFMove(initialValue), WTFMove(promise)));
+        internalObserver->suspendIfNeeded();
+        return internalObserver;
+    }
+
+private:
+    void next(JSC::JSValue value) final
+    {
+        if (!m_accumulator) {
+            m_idx++;
+            m_accumulator = value;
+            return;
+        }
+
+        auto* globalObject = protectedScriptExecutionContext()->globalObject();
+        ASSERT(globalObject);
+
+        Ref vm = globalObject->vm();
+
+        JSC::JSLockHolder lock(vm);
+        auto scope = DECLARE_CATCH_SCOPE(vm);
+
+        auto result = protectedCallback()->handleEventRethrowingException(*m_accumulator, value, m_idx++);
+
+        JSC::Exception* exception = scope.exception();
+        if (UNLIKELY(exception)) {
+            scope.clearException();
+            auto value = exception->value();
+            protectedPromise()->reject<IDLAny>(value);
+            Ref { m_signal }->signalAbort(value);
+        }
+
+        if (result.type() == CallbackResultType::Success)
+            m_accumulator = result.releaseReturnValue();
+    }
+
+    void error(JSC::JSValue value) final
+    {
+        protectedPromise()->reject<IDLAny>(value);
+    }
+
+    void complete() final
+    {
+        InternalObserver::complete();
+
+        if (!m_accumulator) {
+            protectedPromise()->reject(Exception { ExceptionCode::TypeError, "No inital value for Observable with no values"_s });
+            return;
+        }
+
+        protectedPromise()->resolve<IDLAny>(m_accumulator.value());
+    }
+
+    void visitAdditionalChildren(JSC::AbstractSlotVisitor& visitor) const final
+    {
+        protectedCallback()->visitJSFunction(visitor);
+    }
+
+    Ref<DeferredPromise> protectedPromise() const { return m_promise; }
+    Ref<ReducerCallback> protectedCallback() const { return m_callback; }
+
+    InternalObserverReduce(ScriptExecutionContext& context, Ref<AbortSignal>&& signal, Ref<ReducerCallback>&& callback, std::optional<JSC::JSValue>&& initialValue, Ref<DeferredPromise>&& promise)
+        : InternalObserver(context)
+        , m_signal(WTFMove(signal))
+        , m_callback(WTFMove(callback))
+        , m_accumulator(WTFMove(initialValue))
+        , m_promise(WTFMove(promise))
+    {
+    }
+
+    uint64_t m_idx { 0 };
```

Comment:
WebKit coding style calls this `m_index`, not `m_idx`, because we use words instead of abbreviations.